### ISSUE TITLE
Tailwind IntelliSense does not list my custom class in @layer Utilities

### ISSUE THREAD

**Issue by huuphongnguyen** [tailwindlabs/tailwindcss-intellisense#227]
I have created some custom style class inside @Layer utilities {}. And when I used those classes in Classname="...", Intellisense did not recommend or list them. But it still works.

Example in my global.css:
`@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .phong {
    width: 300px;
    height: 300px;
  }
}`

 and I use it: `<div className="phong bg-red-200">`. It's still working but `phong` was not listed by Intellisense when typing.

**Comment by Nagell** [comment]
Anything new about this? It would be really helpful. Without this feature, we have to continuously check other files for our custom classes and 'everything is mixin'-property of tailwind can be barely embraced.

**Comment by happycollision** [comment]
Since @bradlc is closing other issues in favor of this one, it might be good to see this comment from him on issue #285:

> This is kind of related to #227 in that the core issue is that the extension isn't aware of custom CSS at all. This is a non-trivial issue to solve and there's no immediate plans to implement it.

**Comment by 7iomka** [comment]
Maybe someone knows another extension that could do hints (at least just by the name of the class) based on the analysis of all the styles of the open window (workspace / monorepo list of workspaces)?

**Comment by glidelos** [comment]
> Maybe someone knows another extension that could do hints (at least just by the name of the class) based on the analysis of all the styles of the open window (workspace / monorepo list of workspaces)?

Hi! You can use scss everywhere plugin, then click on thunderbolt button in the status bar for first use or refresh cache of style classes and it's done! ;)

**Comment by Nagell** [comment]
> Hi! You can use scss everywhere plugin, then click on thunderbolt button in the status bar for first use or refresh cache of style classes and it's done! ;)

it's not available in .pcss files, so I've just asked if they could add it.
https://github.com/gencer/SCSS-Everywhere/issues/75

**Comment by boompikachu** [comment]
If you add it as a plugin, it will show up. 

```js
const plugin = require("tailwindcss/plugin");

module.exports = {
  content: [],
  theme: {},
  plugins: [
    plugin(function ({ addUtilities }) {
      addUtilities({
        ".phong": {
          width: "300px",
          height: "300px",
        },
      });
    }),
  ],
};
```

**Comment by Nagell** [comment]
C'mon, change it to Feature request, so we can look forward if someone will try to create this feature. It's also a help topic for all newcomers.

**Comment by haddigan** [comment]
I found a workaround, sharing it in case anyone is interested. this works in tailwind 3, not sure about previous versions.

Instead of creating custom components and utilities in your tailwind.css file, you can create a custom plugin and define them in JS instead. Your custom classes will get picked up by Tailwind Intellisense that way.

before: 
```
// tailwind.css
@layer utilities {
  .phong {
    height: 300px;
    width: 300px;
  }
}
```
![image](https://user-images.githubusercontent.com/10161907/170764634-ce1e07c7-6e61-4258-aca1-293ea073bafa.png)


after: 
```
// tailwind.config.js

module.exports = {
  plugins: [
    ({ addUtilities }) => {
      addUtilities({
        '.phong': {
          width: '300px',
          height: '300px',
        }
      });
    }
  ]
}
```
![image](https://user-images.githubusercontent.com/10161907/170764826-16dc091a-5fb6-40dd-af00-24e18ae2c1e5.png)

**Comment by SebHex** [comment]
For anyone who wants to add type to @haddigan's [workaround](https://github.com/tailwindlabs/tailwindcss-intellisense/issues/227#issuecomment-1139895799):

```diff
// tailwind.config.js

+ const plugin = require('tailwindcss/plugin')

module.exports = {
  plugins: [
-   ({ addUtilities }) => {
+   plugin(({ addUtilities }) => {
      addUtilities({
        '.phong': {
          width: '300px',
          height: '300px',
        }
      });
-   }
+   })
  ]
}
```



OR:

```diff
// tailwind.config.js

module.exports = {
  plugins: [
+   /** @type {import('tailwindcss/types/config').PluginCreator} */
    ({ addUtilities }) => {
      addUtilities({
        '.phong': {
          width: '300px',
          height: '300px',
        }
      });
    }
  ]
}
```

Read more at [Tailwind's plugins documentation](https://tailwindcss.com/docs/plugins#adding-utilities).

**Comment by rik** [comment]
Although the workaround can be useful, it makes writing custom CSS more cumbersome given the CSS-in-JS syntax and that `@apply` isn't available anymore.

So I went looking for a complementary extension and found https://marketplace.visualstudio.com/items?itemName=Zignd.html-css-class-completion. The drawback is that this won't display the declarations in the tooltip. On the positive side, you don't have to change your process.

**Comment by wiredearp** [comment]
I just went through several extensions that claim to offser autocomplete for CSS variables, here is the [the winner](https://github.com/willofindie/vscode-cssvar)), and they simply scan the file system to come up with these things. In other words, these extensions *do* know about my source files because I configured that search pattern in `settings.json` myself. I can't begin to imagine how extensions actually work, but source file indexing doesn't appear to be a limitation of the platform. I'd even take a hardcoded list of undiscovered classnames just to get the autocomplete working.

```json
"tailwindCSS.trickyLayerClasses": [
  "one",
  "two",
  "three"
]
```

I tried to see if the plugin workaround could work for me, I had hoped I could declare the classnames in JS while keeping the styling in CSS, but it appears that my entire custom theme gets nuked as soon is I `require('tailwindcss/plugin')` in the code. Upon restart of VSCode, this extension only knows about the native Tailwind classnames. Except that of course is another issue altogether.

**Comment by david-plugge** [comment]
Even though my previous solution works, i often find myself restarting vscode. It seems like node really doesn´t like `require.extensions['.css']`...
To avoid that i created [tailwindcss-extend](https://github.com/david-plugge/tailwindcss-extend) that comes with a cli aswell as a vite plugin to compile your components into a tailwind plugin. I´m happy about some feedback!

**Comment by Senbonzakura1234** [comment]
> Even though my previous solution works, i often find myself restarting vscode. It seems like node really doesn´t like `require.extensions['.css']`... To avoid that i created [tailwindcss-extend](https://github.com/david-plugge/tailwindcss-extend) that comes with a cli aswell as a vite plugin to compile your components into a tailwind plugin. I´m happy about some feedback!

Can i install this as dev-dependencies? how should i use this with nextjs?

**Comment by padcom** [comment]
I was just trying to add myself a new utility class to the css and I was surprised to learn that vscode has no clue about those classes. It looks like this ticket is closed but the issue described is still unresolved. What shall we do? Is the plugin-way the only option that will ever be implemented? Can we have that baked-in to vite instead of writing it manually for every project?

**Comment by david-plugge** [comment]
@padcom there was a [talk](https://youtu.be/CLkxRnRQtDE?t=3226) from the tailwind authors about the future of tailwind. A new tool called [lightningcss](https://lightningcss.dev/) is eventually going to be used which opens up a so many possibilities dues to the performance of rust. 

But for now this is probably the only way to get it working.

**Comment by david-plugge** [comment]
> Can i install this as dev-dependencies? how should i use this with nextjs?

Save it as a dev-dependency and use `npm-run-all` or a similar tool to start the next server aswell as tailwindcss-extend

(i didn´t test this)
```json
{
  "scripts": {
    "dev": "run-p dev:*", // run all scripts starting with 'dev:' in parallel
    "dev:tailwindcss-extend": "tailwindcss-extend -w -p **/styles/*.css",
    "dev:next": "whatever you need to start next"
  },
  "devDependencies": {
    "npm-run-all": "latest",
    "tailwindcss-extend": "latest"
  }
}
```

**Comment by theo-bittencourt** [comment]
I'm on a "Nodeless" stack with Tailwind Standalone CLI , so only first-party plugins are supported.

david-plugge's solution is not viable.

Only `addUtilities({})` approach seems to work.

**Comment by pabloharbar** [comment]
I added recursivity to the plugin based on @david-plugge 's answer.

``` javascript
const postcss = require("postcss");
const postcssJs = require("postcss-js");
const { readFileSync } = require("fs");
const path = require("path");

function handle_css_file(filename, addBase, addComponents, addUtilities) {
    const css = readFileSync(filename, "utf8");
    const root = postcss.parse(css);
    const jss = postcssJs.objectify(root);

    if ("@layer base" in jss) {
        addBase(jss["@layer base"]);
    }
    if ("@layer components" in jss) {
        addComponents(jss["@layer components"]);
    }
    if ("@layer utilities" in jss) {
        addUtilities(jss["@layer utilities"]);
    }
    for (const key in jss) {
        if (key.includes("@import")) {
            const match = key.match(/"([^"]+)"/);
            if (match && !key.includes("tailwindcss")) {
                const importedFile = match[1];
                const file_path = path.resolve(path.dirname(filename), importedFile);
                handle_css_file(file_path, addBase, addComponents, addUtilities);
            }
        }
    }
}

require.extensions[".css"] = function (module, filename) {
    module.exports = ({ addBase, addComponents, addUtilities }) => {
        handle_css_file(filename, addBase, addComponents, addUtilities);
    };
};

```

This comes in handy when tailwind entry point has import statements such as:

``` javascript
@import "tailwindcss/base";
@import "tailwindcss/components";
@import "tailwindcss/utilities";

/* Custom overriding styles */
@import "./style/text.css";
```

In this case the recursive plugin handles the import statement files.

**Comment by Milkywayrules** [comment]
> For people coming to this thread, I've also settled with relying on plugin for intellisense of custom class names. And for what it's worth, you can do `@apply` in JS:
> 
> ```diff
> // tailwind.config.js
> 
> module.exports = {
>   plugins: [
>     ({ addUtilities }) => {
>       addUtilities({
>         '.phong': {
> +         '@apply w-[300px] h-[300px]': {},
> -         width: '300px',
> -         height: '300px',
>         }
>       });
>     }
>   ]
> }
> ```

oh yeah! this is works and really good, this works very well every I save my tw config too 👍 thanks, mate! @vnphanquang

**Comment by alpha-mo** [comment]
I don't know about you guys but it seems like an obvious feature to me, no one wants to write css in js files (in the case of using addUtilities) 
```css
@layer utilities{
  .my-class{
    /* my css which can be tidious to write in js */
  }
}
```
```js
addUtilities({'my-class': ".my-class"})
```
or even:
```js
addUtilities({'my-class'})
```

**Comment by nuhptr** [comment]
cmon tailwind team, this feature help a lot @adamwathan

**Comment by gangsthub** [comment]
I found this also happens when you use `addBase` on a Tailwind plugin ([docs](https://tailwindcss.com/docs/plugins#adding-base-styles)). The classes added via `addBase` are not listed in the IDE autocompletion.

Maybe because, as said in the docs, `addBase` is not meant to be used with classes:

> base styles are meant to target bare selectors like div or h1

IMO, the docs should be clearer about classes not being listed in the `base` layer or this could be treated as a bug, @bradlc.

**Comment by gitmibrahim** [comment]
I also found this working as in scss:
```js
plugins: [
  ({ addUtilities }) => {
    addUtilities({
      '.line-height': {
        '&-5': {
          lineHeight: '1.25rem'
        },
        '&-6': {
          lineHeight: '1.5rem'
        }
      }
    });
  }
]
```

**Comment by G3EORG3E** [comment]
Why is this closed when it still does not work...

**Comment by daltonmenezes** [comment]
For those still experiencing this issue in a monorepo, the solution is:

**.vscode/settings.json**
```json
"tailwindCSS.experimental.configFile": {
    "apps/desktop/tailwind.config.ts": "apps/desktop/**",
    "apps/web/tailwind.config.ts": "apps/web/**"
  },
```
now, the tailwindcss extension will be able to look at your config files properly!

**Comment by seancheung** [comment]
`require.extensions` is deprecated. Use a custom plugin-creator function instead:

cssplugin.js
```js
// @ts-check

import { parse } from 'postcss'
import { objectify } from 'postcss-js'
import { readFileSync } from 'fs'

/**
 * CSS as plugin
 * @param {string} filename CSS file
 * @returns {import('tailwindcss/types/config').PluginCreator}
 */
export default filename => {
  return ({ addBase, addComponents, addUtilities }) => {
    const css = readFileSync(filename, 'utf8')
    const root = parse(css)
    const jss = objectify(root)

    if ('@layer base' in jss) {
      addBase(jss['@layer base'])
    }
    if ('@layer components' in jss) {
      addComponents(jss['@layer components'])
    }
    if ('@layer utilities' in jss) {
      addUtilities(jss['@layer utilities'])
    }
  }
}
```

cssplugin.d.ts
```ts
import { PluginCreator } from 'tailwindcss/types/config'

declare const e: (filename: string) => PluginCreator

export = e
```

tailwind.config.js
```js
import { resolve } from 'node:path'
import cssplugin from './cssplugin'

/** @type {import('tailwindcss').Config} */
export default {
    plugins: [cssplugin(resolve(__dirname, './assets/myUtilities.css'))],
}
```

**Comment by ProkhoDim** [comment]
> `require.extensions` is deprecated. Use a custom plugin-creator function instead:
> 
> cssplugin.js
> 
> ```js
> // @ts-check
> 
> import { parse } from 'postcss'
> import { objectify } from 'postcss-js'
> import { readFileSync } from 'fs'
> 
> /**
>  * CSS as plugin
>  * @param {string} filename CSS file
>  * @returns {import('tailwindcss/types/config').PluginCreator}
>  */
> export default filename => {
>   return ({ addBase, addComponents, addUtilities }) => {
>     const css = readFileSync(filename, 'utf8')
>     const root = parse(css)
>     const jss = objectify(root)
> 
>     if ('@layer base' in jss) {
>       addBase(jss['@layer base'])
>     }
>     if ('@layer components' in jss) {
>       addComponents(jss['@layer components'])
>     }
>     if ('@layer utilities' in jss) {
>       addUtilities(jss['@layer utilities'])
>     }
>   }
> }
> ```
> 
> cssplugin.d.ts
> 
> ```ts
> import { PluginCreator } from 'tailwindcss/types/config'
> 
> declare const e: (filename: string) => PluginCreator
> 
> export = e
> ```
> 
> tailwind.config.js
> 
> ```js
> import { resolve } from 'node:path'
> import cssplugin from './cssplugin'
> 
> /** @type {import('tailwindcss').Config} */
> export default {
>     plugins: [cssplugin(resolve(__dirname, './assets/my-utilities.css'))],
> }
> ```

If you use tailwindcss with nextjs turbopack probably you will have next error:
```
lint TP1004 fs.readFileSync(???*0*, "utf8") is very dynamic
  13 | export default function cssAsPlugin(filename) {
  14 |   return ({ addBase, addComponents, addUtilities }) => {
> 15 |     const css = readFileSync(filename, 'utf8');
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  16 |     const root = parse(css);
  17 |     const jss = objectify(root);
  18 |
  ```
  
For me the solution to have autocomplete without such errors was to hardcode the filenames (I think, that if you have such functions you don't need a 'dynamic' usage)
  
  So my code looks like:
  ```
  /**
 * CSS as plugin
 * @param {string} filename CSS file
 * @returns {import('tailwindcss/types/config').PluginCreator[]}
 */
export default function cssAsPlugin() {
  const css_files = [
    readFileSync(resolve(__dirname, './src/app/styles/tailwind/base.css'), 'utf8'),
    readFileSync(resolve(__dirname, './src/app/styles/tailwind/components.css'), 'utf8'),
    readFileSync(resolve(__dirname, './src/app/styles/tailwind/utilities.css'), 'utf8'),
  ];
  return css_files.map((css) => ({ addBase, addComponents, addUtilities }) => {
    const root = parse(css);
    const jss = objectify(root);

    if ('@layer base' in jss) {
      addBase(jss['@layer base']);
    }
    if ('@layer components' in jss) {
      addComponents(jss['@layer components']);
    }
    if ('@layer utilities' in jss) {
      addUtilities(jss['@layer utilities']);
    }
  });
}
  ```
  
  Also if you import this files in root css file you will get a duplicate styles. For example:
  
  1. You have file called `components.css` in which you defined layer `.primary-button`
  2. You imported this file in `index.css` and add it to function `cssAsPlugin`
  3. If you look in code inspector you will see that there is two same `.primary-button` classes.
  4. Just not import this file in `index.css` and you will se only one class

**Comment by reorx** [comment]
Please reopen the feature request. It should remain open unless you explicitly state that it is not being considered.

**Comment by gesshoo** [comment]
> For those still experiencing this issue in a monorepo, the solution is:
> 
> **.vscode/settings.json**
> 
> ```json
> "tailwindCSS.experimental.configFile": {
>     "apps/desktop/tailwind.config.ts": "apps/desktop/**",
>     "apps/web/tailwind.config.ts": "apps/web/**"
>   },
> ```
> 
> now, the tailwindcss extension will be able to look at your config files properly!

This worked for me!

You need no edit the json file in the Tailwind CSS Intellisense extension. ( Tailwind CSS > Experimental: Config File - Edit in settings.json )

Since I'm using Laravel (macOS Catalina) this is my setup:

I changed this: 
"tailwindCSS.experimental.configFile": null

To this:
"tailwindCSS.experimental.configFile": {
    "tailwind.config.js": "app/**",
    "tailwind.config.js": "resources/**"
  },

**Comment by HarshArora-1205** [comment]
> You need no edit the json file in the Tailwind CSS Intellisense extension. ( Tailwind CSS > Experimental: Config File - Edit in settings.json )

When I change my settings.json, it stopped suggesting tailwind classes itself. :/

Is there any other solution, without the hassle of specifying each custom utility class in addutilities()?

**Comment by mCzolko** [comment]
@seancheung solutions works like a charm

**Comment by karlhorky** [comment]
> @bradlc in [comment 1046763137](https://github.com/tailwindlabs/tailwindcss-intellisense/issues/227#issuecomment-1046763137) (Feb 2022): I agree it would be cool if we could get custom classes to appear in the completion list! As I said in #285 this is non-trivial, because the extension would need to know about your source files and how to compile them (if necessary). Maybe we can look into it in the future. As this is a feature request and not a bug I am going to close this issue.

@thecrypticace @bradlc does Brad's comment above still apply here now with the advent of Tailwind CSS v4 and the CSS configuration API?

Given the solutions like [the plugin](https://github.com/tailwindlabs/tailwindcss-intellisense/issues/227#issuecomment-2344596934) by @seancheung that are appearing in this thread, does it make sense to codify this (or maybe some better solution with the CSS configuration API) into either of the following?

1. A new 1st-party feature for the VS Code extension
2. A blessed, documented, 3rd party option

If any of these paths forward would be considered, would it be better to reopen this issue? Or rather create a new issue with only the new approach and then link it?

**Comment by thecrypticace** [comment]
Nah, it does not. Classes defined `.css` files show up when using v4. 👍 

The syntax for that is different in v4 btw, `@utility …` so we're not hijacking CSS native layers any more.

There are still significant complications like bundlers knowing about tsconfig paths that we *also* have to know about (I'm literally working on this right now as I type this). But in general it works.

**Comment by JoaoTeixeira20** [comment]
> I managed to get it working. It makes use of `require.extensions` which is deprecated (for good reason) but that is the only way i can get vscode and vite to auto-reload the tailwind config when the css file changes.
> 
> The helper creates a seperate tailwind plugin for every css file we `require`.
> 
> // cssAsPlugin.js
> const postcss = require('postcss');
> const postcssJs = require('postcss-js');
> const { readFileSync } = require('fs');
> 
> require.extensions['.css'] = function (module, filename) {
> 	module.exports = ({ addBase, addComponents, addUtilities }) => {
> 		const css = readFileSync(filename, 'utf8');
> 		const root = postcss.parse(css);
> 		const jss = postcssJs.objectify(root);
> 
> 		if ('@layer base' in jss) {
> 			addBase(jss['@layer base']);
> 		}
> 		if ('@layer components' in jss) {
> 			addComponents(jss['@layer components']);
> 		}
> 		if ('@layer utilities' in jss) {
> 			addUtilities(jss['@layer utilities']);
> 		}
> 	};
> };
> // tailwind.config.js
> require('./cssAsPlugin.cjs');
> 
> /** @type {import('tailwindcss').Config} */
> const config = {
> 	content: ['./src/**/*.{html,js,svelte,ts}'],
> 
> 	theme: {
> 		extend: {}
> 	},
> 
> 	plugins: [require('./components.css')]
> };
> 
> module.exports = config;
> /* components.css */
> @layer components {
> 	.btn {
> 		@apply mx-2 rounded-lg bg-blue-500 px-4 py-2 text-white;
> 	}
> 
> 	.error {
> 		color: theme(colors.red.500);
> 	}
> }

to avoid using require.extensions i made a slightly change on this:

//cssAsPlugin.js
`import fs from "fs";
import path from "path";
import postcss from "postcss";
import postcssJs from "postcss-js";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const cssPath = path.resolve(__dirname, "./src/styles/theme.css");
console.log(cssPath);
const css = fs.readFileSync(cssPath, "utf8");
const jss = postcssJs.objectify(postcss.parse(css));

export default function cssAsPlugin({ addUtilities, addComponents, addBase }) {
  if (jss["@layer utilities"]) addUtilities(jss["@layer utilities"]);
  if (jss["@layer components"]) addComponents(jss["@layer components"]);
  if (jss["@layer base"]) addBase(jss["@layer base"]);
}
`

//tailwind.config.ts
`/** @type {import('tailwindcss').Config} */
import cssAsPluginfrom './components.js';

module.exports = {
  content: [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
  ],
  theme: {
    extend: {},
  },
  plugins: [cssAsPlugin],
};
`

//theme.css
`@layer utilities {
  .theme-card {
    @apply p-1 md:p-2 lg:p-3 xl:p-4;
  }

  .theme-layout {
    @apply p-2 md:p-4 lg:p-6 xl:p-8;
  }
}
`

classes work along with intellisense
